tests/controller_tests/test_mappings.py

```diff
@@ -0,0 +1,186 @@
+    mapping = DataMapping("worksheet_name", MockModel, column_mapping)  # noqa
+
+    # foreign key relation
+    fk_mapping = FKMapping(
+        parent_lookup="parent_lookup", parent_model=MockParentModel, child_fk="fk", child_lookup="fk_lookup_col"
+    )
+    mapping.fk_relations.append(fk_mapping)
+
+    models = mapping.map_worksheet_to_models(worksheet)
+
+    # mapping returns the correct number of results
+    assert len(models) == 2
+
+    # FK attribute has the correct mocked value from mocked_get_row_id
+    assert models[0].fk == "123"
+    assert models[1].fk == "123"
+
+    # the other model attributes have been mapped correctly
+    assert models[0].db_col1 == "A"
+    assert models[0].db_col2 == 1
+    assert models[1].db_col1 == "B"
+    assert models[1].db_col2 == 2
+
+
+def test_data_mapping_project_id_fk(mocked_get_row_id):
+    """Test the special path for when we map a project_id FK. This requires the model to also include submission_id."""
+    # sample worksheet and mapping
+    worksheet = pd.DataFrame(
+        [
+            {"fk_lookup_col": "lookup1", "submission_id": "sub_id_value"},
+        ]
+    )
+    column_mapping = {"fk_lookup_col": "fk_lookup_col", "submission_id": "submission_id"}
+    mapping = DataMapping("worksheet_name", MockModel, column_mapping)  # noqa
+
+    # project_id foreign key relation
+    fk_mapping = FKMapping(
+        parent_lookup="parent_lookup", parent_model=MockParentModel, child_fk="project_id", child_lookup="fk_lookup_col"
+    )
+    mapping.fk_relations.append(fk_mapping)
+
+    models = mapping.map_worksheet_to_models(worksheet)
+
+    assert len(models) == 1
+    assert models[0].project_id == "123"  # mocked_get_row_id return value
+    assert models[0].submission_id == "sub_id_value"
+
+
+def test_data_mapping_child_fk_attribute_matches_lookup_attribute(mocked_get_row_id):
+    """Test the special path for when a child fk attribute name is the same as the lookup attribute.
+    In this case, do don't want to delete the child attribute lookup because it's value has been replaced with the FK.
+    """
+    # sample worksheet and mapping
+    worksheet = pd.DataFrame(
+        [
+            {"child_fk_and_lookup": "lookup1"},
+        ]
+    )
+    column_mapping = {
+        "child_fk_and_lookup": "child_fk_and_lookup",
+    }
+    mapping = DataMapping("worksheet_name", MockModel, column_mapping)  # noqa
+
+    # project_id foreign key relation
+    fk_mapping = FKMapping(
+        parent_lookup="parent_lookup",
+        parent_model=MockParentModel,
+        child_fk="child_fk_and_lookup",
+        child_lookup="child_fk_and_lookup",
+    )
+    mapping.fk_relations.append(fk_mapping)
+
+    models = mapping.map_worksheet_to_models(worksheet)
+
+    assert len(models) == 1
+    assert models[0].child_fk_and_lookup == "123"  # mocked_get_row_id return value
+
+
+def test_get_row_id_row_found(mock_db_session):
```

Comment:
I'm not sure these last 3 tests are strictly necessary, since they are testing a 1 line function with all it's parts mocked (so it effectively is just testing that it returns the mock calls you said it should immediately above).

But happy for them to stay in if you want, I guess they help demonstrate the expected use-cases of the function

Reply:
It isn't quite a one line function but yeah I agree, it isn't testing much. I was struggling with how best to test that function. Once I was done with wrestling with mocks I came to the same conclusion. 

Do you have an ideas of how to improve it? It still tests the various scenarios after the mocked query results have been returned.

Reply:
nah, it's probably fine if it's actually testing scenarios then I'm happy.